### PR TITLE
Refactor VoiceCommandManager for taskmaster

### LINKED ISSUE #1549: Refactor VoiceCommandManager to use queues
### Bug Report
Other Screen Manager managers use operation queues to manage updates in a manageable way. We should do the same for the `VoiceCommandManager`.

### PR DESCRIPTION
**Pull request by RHenigan**
Fixes #1549 

This PR is **[ready]** for review.

### Risk
This PR makes **[minor]** API changes.

### Testing Plan
- [X] I have verified that I have not introduced new warnings in this PR (or explain why below)
- [X] I have run the unit tests with this PR
- [X] I have tested this PR against Core and verified behavior (if applicable, if not applicable, explain why below).
- [X] I have tested Android, Java SE, and Java EE

#### Unit Tests
Updated VoiceCommandManagerTests to accound for taskmaster dependancy'

#### Core Tests
In the HelloSdl App, update your `setVoiceCommands()` method to set the voiceCommands every 5 seconds (example below). every few seconds the voiceCommands list should change  between the two lists. Let this run for a few minutes and verify there are no issues setting the voiceCommands.

```
Timer timer = new Timer();
        timer.schedule(new TimerTask() {
            boolean firstGroup = true;
            @Override
            public void run() {
                if (firstGroup) {
                    sdlManager.getScreenManager().setVoiceCommands(Arrays.asList(voiceCommand1, voiceCommand2));
                    firstGroup = false;
                } else {
                    sdlManager.getScreenManager().setVoiceCommands(Arrays.asList(voiceCommand3, voiceCommand4));
                    firstGroup = true;
                }

            }
        }, 0, 5000);
```

### Summary
This PR will create `VoiceCommandReplaceOperation` which will use send the list of `DeleteCommand` and `AddCommand` RPCs. `BaseVoiceCommandManager` has been refactored to use taskMaster Queues.

### CLA
- [X] I have signed [the CLA](https://docs.google.com/forms/d/e/1FAIpQLSdsgJY33VByaX482zHzi-xUm49JNnmuJOyAM6uegPQ2LXYVfA/viewform)

### CODE REVIEW

**Comment by RHenigan** [comment]
[(5) Refactor `VoiceCommandManager` to use queues [1]](https://trello.com/c/jyGUBXpC/1562-5-refactor-voicecommandmanager-to-use-queues-1)

**Review by bilal-alsharifi — CHANGES_REQUESTED** [review]

File: base/src/main/java/com/smartdevicelink/managers/screen/menu/BaseVoiceCommandManager.java [hunk 1]
```diff
@@ -141,135 +164,40 @@ public void setVoiceCommands(List<VoiceCommand> voiceCommands) {
...
-                if (listener != null) {
-                    listener.onComplete(true);
-                }
-            }
-
-            @Override
-            public void onResponse(int correlationId, RPCResponse response) {
+            public void updateVoiceCommands(List<VoiceCommand> newCurrentVoiceCommands, HashMap<RPCRequest, String> errorObject) {
+                DebugTool.logInfo(TAG, "The updated list of VoiceCommands: " + newCurrentVoiceCommands);
+                DebugTool.logError(TAG, "The failed Add and Delete Commands: " + errorObject);
```
[the remark this answers is not in the document]
Comment (RHenigan) [reply base/src/main/java/com/smartdevicelink/managers/screen/menu/BaseVoiceCommandManager.java]: Updated to only print when errorObject is not empty

File: base/src/main/java/com/smartdevicelink/managers/screen/menu/VoiceCommandUpdateOperation.java [hunk 2]
```diff
@@ -0,0 +1,228 @@
...
+import com.smartdevicelink.proxy.rpc.DeleteCommandResponse;
+import com.smartdevicelink.proxy.rpc.listeners.OnMultipleRequestListener;
+import com.smartdevicelink.util.DebugTool;
+
+import java.lang.ref.WeakReference;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+
+public class VoiceCommandUpdateOperation extends Task {
```
[the remark this answers is not in the document]
Comment (RHenigan) [reply base/src/main/java/com/smartdevicelink/managers/screen/menu/VoiceCommandUpdateOperation.java]: fixed

File: base/src/main/java/com/smartdevicelink/managers/screen/menu/BaseVoiceCommandManager.java [hunk 3]
```diff
@@ -90,12 +93,15 @@ public void dispose() {
...
         voiceCommands = null;
-        oldVoiceCommands = null;
+        currentVoiceCommands = null;
 
         waitingOnHMIUpdate = false;
         currentHMILevel = null;
-        inProgressUpdate = null;
-        hasQueuedUpdate = false;
+
+        transactionQueue.close();
```
[the remark this answers is not in the document]
Comment (RHenigan) [reply base/src/main/java/com/smartdevicelink/managers/screen/menu/BaseVoiceCommandManager.java]: added null check

File: base/src/main/java/com/smartdevicelink/managers/screen/menu/BaseVoiceCommandManager.java [hunk 4]
```diff
@@ -104,6 +110,23 @@ public void dispose() {
...
+            transactionQueue.pause();
+        } else {
+            DebugTool.logInfo(TAG, "Starting the transaction queue");
+            transactionQueue.resume();
+        }
+    }
+
     // SETTERS
 
     public void setVoiceCommands(List<VoiceCommand> voiceCommands) {
```
[the remark this answers is not in the document]
Comment (RHenigan) [reply base/src/main/java/com/smartdevicelink/managers/screen/menu/BaseVoiceCommandManager.java:134]: Updated to allow for an emptylist to be set, added unit test for the updateOperation

File: base/src/main/java/com/smartdevicelink/managers/screen/menu/BaseVoiceCommandManager.java [hunk 5]
```diff
@@ -125,9 +148,9 @@ public void setVoiceCommands(List<VoiceCommand> voiceCommands) {
         waitingOnHMIUpdate = false;
```
[the remark this answers is not in the document]
Comment (RHenigan) [reply base/src/main/java/com/smartdevicelink/managers/screen/menu/BaseVoiceCommandManager.java]: removed old HMI Status logic

File: base/src/main/java/com/smartdevicelink/managers/screen/menu/BaseVoiceCommandManager.java [hunk 6]
```diff
@@ -34,47 +34,50 @@
...
     private static final String TAG = "BaseVoiceCommandManager";
-    List<VoiceCommand> voiceCommands, oldVoiceCommands;
-
-    List<AddCommand> inProgressUpdate;
+    List<VoiceCommand> voiceCommands, currentVoiceCommands;
 
     int lastVoiceCommandId;
     private static final int voiceCommandIdMin = 1900000000;
 
     boolean waitingOnHMIUpdate;
```
[the remark this answers is not in the document]
Comment (RHenigan) [reply base/src/main/java/com/smartdevicelink/managers/screen/menu/BaseVoiceCommandManager.java]: removed old HMI Status Logic

**Review by bilal-alsharifi — CHANGES_REQUESTED** [review]

File: base/src/main/java/com/smartdevicelink/managers/screen/menu/BaseVoiceCommandManager.java [hunk 7]
```diff
@@ -104,30 +109,38 @@ public void dispose() {
...
+            transactionQueue.pause();
+        } else {
+            DebugTool.logInfo(TAG, "Starting the transaction queue");
+            transactionQueue.resume();
+        }
+    }
+
     // SETTERS
 
     public void setVoiceCommands(List<VoiceCommand> voiceCommands) {
```
[the remark this answers is not in the document]
Comment (RHenigan) [reply base/src/main/java/com/smartdevicelink/managers/screen/menu/BaseVoiceCommandManager.java:131]: added a check

File: base/src/main/java/com/smartdevicelink/managers/screen/menu/BaseVoiceCommandManager.java [hunk 8]
```diff
@@ -104,30 +109,38 @@ public void dispose() {
...
-        // make sure hmi is not none
-        if (currentHMILevel == null || currentHMILevel == HMILevel.HMI_NONE) {
-            // Trying to send on HMI_NONE, waiting for full
-            this.voiceCommands = new ArrayList<>(voiceCommands);
-            waitingOnHMIUpdate = true;
-            return;
-        }
-
-        waitingOnHMIUpdate = false;
         lastVoiceCommandId = voiceCommandIdMin;
```
[the remark this answers is not in the document]
Comment (RHenigan) [reply base/src/main/java/com/smartdevicelink/managers/screen/menu/BaseVoiceCommandManager.java]: removed this assignment

File: base/src/main/java/com/smartdevicelink/managers/screen/menu/BaseVoiceCommandManager.java [hunk 9]
```diff
@@ -104,30 +109,38 @@ public void dispose() {
...
         updateIdsOnVoiceCommands(voiceCommands);
-        this.oldVoiceCommands = new ArrayList<>();
-        if (this.voiceCommands != null && !this.voiceCommands.isEmpty()) {
-            this.oldVoiceCommands.addAll(this.voiceCommands);
+        this.currentVoiceCommands = new ArrayList<>();
```
[the remark this answers is not in the document]
Comment (RHenigan) [reply base/src/main/java/com/smartdevicelink/managers/screen/menu/BaseVoiceCommandManager.java]: removed this to align with iOS code

File: base/src/main/java/com/smartdevicelink/managers/screen/menu/BaseVoiceCommandManager.java [hunk 10]
```diff
@@ -141,135 +154,42 @@ public void setVoiceCommands(List<VoiceCommand> voiceCommands) {
     // UPDATING SYSTEM
 
     private void update() {
```
[the remark this answers is not in the document]
Comment (RHenigan) [reply base/src/main/java/com/smartdevicelink/managers/screen/menu/BaseVoiceCommandManager.java]: removed update method and added to set voice commands

File: base/src/main/java/com/smartdevicelink/managers/screen/menu/BaseVoiceCommandManager.java [hunk 11]
```diff
@@ -141,135 +154,42 @@ public void setVoiceCommands(List<VoiceCommand> voiceCommands) {
...
-        if (voiceCommands == null || voiceCommands.size() == 0) {
-            if (listener != null) {
-                listener.onComplete(true); // no voice commands to send doesnt mean that its an error
-            }
-            return;
+    private void cleanTransactionQueue() {
+        if (transactionQueue.getTasksAsList().size() > 0) {
+            updateOperation = null;
+            transactionQueue.clear();
         }
```
[the remark this answers is not in the document]
Comment (RHenigan) [reply base/src/main/java/com/smartdevicelink/managers/screen/menu/BaseVoiceCommandManager.java]: changed to null check

File: base/src/main/java/com/smartdevicelink/managers/screen/menu/BaseVoiceCommandManager.java [hunk 12]
```diff
@@ -141,135 +154,42 @@ public void setVoiceCommands(List<VoiceCommand> voiceCommands) {
...
-    // DELETES
-
-    List<DeleteCommand> deleteCommandsForVoiceCommands(List<VoiceCommand> voiceCommands) {
-        List<DeleteCommand> deleteCommandList = new ArrayList<>();
-        for (VoiceCommand command : voiceCommands) {
-            DeleteCommand delete = new DeleteCommand(command.getCommandId());
-            deleteCommandList.add(delete);
+        if (updateOperation != null) {
+            updateOperation.cancelTask();
+            updateOperation = null;
```
[the remark this answers is not in the document]
Comment (RHenigan) [reply base/src/main/java/com/smartdevicelink/managers/screen/menu/BaseVoiceCommandManager.java:169]: removed assignment in the transactionQueue check

File: base/src/main/java/com/smartdevicelink/managers/screen/menu/VoiceCommandUpdateOperation.java [hunk 13]
```diff
@@ -0,0 +1,227 @@
...
+    interface VoiceCommandChangesListener {
+        void updateVoiceCommands(List<VoiceCommand> newCurrentVoiceCommands, HashMap<RPCRequest, String> errorObject);
+    }
+
+    VoiceCommandUpdateOperation(ISdl internalInterface, List<VoiceCommand> oldVoiceCommands, List<VoiceCommand> pendingVoiceCommands, VoiceCommandChangesListener voiceCommandListener) {
+        super("VoiceCommandReplaceOperation");
+        this.internalInterface = new WeakReference<>(internalInterface);
+        this.oldVoiceCommands = oldVoiceCommands;
+        this.pendingVoiceCommands = pendingVoiceCommands;
+        this.currentVoiceCommands = oldVoiceCommands;
```
[the remark this answers is not in the document]
Comment (RHenigan) [reply base/src/main/java/com/smartdevicelink/managers/screen/menu/VoiceCommandUpdateOperation.java]: changed to copy oldVoiceCommands into currentVoiceCommands

File: base/src/main/java/com/smartdevicelink/managers/screen/menu/VoiceCommandUpdateOperation.java [hunk 14]
```diff
@@ -0,0 +1,227 @@
...
+                    @Override
+                    public void onComplete(boolean success2) {
+                        if (!success2) {
+                            DebugTool.logError(TAG, "Error sending voice commands");
+                            onError();
+                        } else {
+                            DebugTool.logInfo(TAG, "Successfully send voice commands");
+                            onFinished();
+                        }
+                        voiceCommandListener.updateVoiceCommands(currentVoiceCommands, errorObject);
```
[the remark this answers is not in the document]
Comment (RHenigan) [reply base/src/main/java/com/smartdevicelink/managers/screen/menu/VoiceCommandUpdateOperation.java]: added null check for listener

File: base/src/main/java/com/smartdevicelink/managers/screen/menu/VoiceCommandUpdateOperation.java [hunk 15]
```diff
@@ -0,0 +1,227 @@
...
+                        DebugTool.logInfo(TAG, "Unable to deleted some old voice commands");
+                        listener.onComplete(false);
+                    }
+                }
+            }
+
+            @Override
+            public void onResponse(int correlationId, RPCResponse response) {
+                DeleteCommandResponse deleteResponse = (DeleteCommandResponse) response;
+                if (!deleteResponse.getSuccess()) {
```
[the remark this answers is not in the document]
Comment (RHenigan) [reply base/src/main/java/com/smartdevicelink/managers/screen/menu/VoiceCommandUpdateOperation.java]: updated

File: base/src/main/java/com/smartdevicelink/managers/screen/menu/VoiceCommandUpdateOperation.java [hunk 16]
```diff
@@ -0,0 +1,227 @@
...
+                    for (DeleteCommand deleteCommand : deleteVoiceCommands) {
+                        if (correlationId == deleteCommand.getCorrelationID()) {
+                            errorObject.put(deleteCommand, response.getInfo());
```
[the remark this answers is not in the document]
Comment (RHenigan) [reply base/src/main/java/com/smartdevicelink/managers/screen/menu/VoiceCommandUpdateOperation.java]: change no longer needed due to change in the onResponse methods

File: base/src/main/java/com/smartdevicelink/managers/screen/menu/VoiceCommandUpdateOperation.java [hunk 17]
```diff
@@ -0,0 +1,227 @@
...
+        for (VoiceCommand command : voiceCommands) {
+            DeleteCommand delete = new DeleteCommand(command.getCommandId());
+            deleteCommandList.add(delete);
+        }
+        return deleteCommandList;
+    }
+
+    private void removeCurrentVoiceCommandWithCorrelationId(int correlationId) {
+        for (DeleteCommand deleteCommand : deleteVoiceCommands) {
+            if (correlationId == deleteCommand.getCorrelationID()) {
```
[the remark this answers is not in the document]
Comment (RHenigan) [reply base/src/main/java/com/smartdevicelink/managers/screen/menu/VoiceCommandUpdateOperation.java]: Updated

File: base/src/main/java/com/smartdevicelink/managers/screen/menu/VoiceCommandUpdateOperation.java [hunk 18]
```diff
@@ -0,0 +1,227 @@
...
+            public void onResponse(int correlationId, RPCResponse response) {
+                DeleteCommandResponse deleteResponse = (DeleteCommandResponse) response;
+                if (!deleteResponse.getSuccess()) {
+                    for (DeleteCommand deleteCommand : deleteVoiceCommands) {
+                        if (correlationId == deleteCommand.getCorrelationID()) {
+                            errorObject.put(deleteCommand, response.getInfo());
+                        }
+                    }
+                } else {
+                    removeCurrentVoiceCommandWithCorrelationId(correlationId);
```
[the remark this answers is not in the document]
Comment (RHenigan) [reply base/src/main/java/com/smartdevicelink/managers/screen/menu/VoiceCommandUpdateOperation.java]: Changed as suggested, new method names are `removeCurrentVoiceCommandForCorrelatingDeleteCommand` and `pendingVoiceCommandForCorrelatingAddCommand`

File: base/src/main/java/com/smartdevicelink/managers/screen/menu/VoiceCommandUpdateOperation.java [hunk 19]
```diff
@@ -0,0 +1,227 @@
...
+                        DebugTool.logInfo(TAG, "Sending Voice Commands Complete with errors");
+                        listener.onComplete(false);
+                    }
+                }
+            }
+
+            @Override
+            public void onResponse(int correlationId, RPCResponse response) {
+                AddCommandResponse addResponse = (AddCommandResponse) response;
+                if (!addResponse.getSuccess()) {
```
[the remark this answers is not in the document]
Comment (RHenigan) [reply base/src/main/java/com/smartdevicelink/managers/screen/menu/VoiceCommandUpdateOperation.java]: updated

File: base/src/main/java/com/smartdevicelink/managers/screen/menu/VoiceCommandUpdateOperation.java [hunk 20]
```diff
@@ -0,0 +1,227 @@
...
+                    for (AddCommand addCommand : addCommandsToSend) {
+                        if (correlationId == addCommand.getCorrelationID()) {
+                            errorObject.put(addCommand, response.getInfo());
```
[the remark this answers is not in the document]
Comment (RHenigan) [reply base/src/main/java/com/smartdevicelink/managers/screen/menu/VoiceCommandUpdateOperation.java]: change no longer needed with changes to onResponse method

File: base/src/main/java/com/smartdevicelink/managers/screen/menu/VoiceCommandUpdateOperation.java [hunk 21]
```diff
@@ -0,0 +1,227 @@
...
+                    onFinished();
+                    return;
+                }
+                // we don't care about errors from deleting, send new add commands
+                sendCurrentVoiceCommands(new CompletionListener() {
+                    @Override
+                    public void onComplete(boolean success2) {
+                        if (!success2) {
+                            DebugTool.logError(TAG, "Error sending voice commands");
+                            onError();
```
[the remark this answers is not in the document]
Comment (RHenigan) [reply base/src/main/java/com/smartdevicelink/managers/screen/menu/VoiceCommandUpdateOperation.java]: Changed to `onFinished()` as this is how it was implemented for other tasks such as `SoftButtonReplaceOperation` and `TextAndGraphicUpdateOperation`

File: base/src/main/java/com/smartdevicelink/managers/screen/menu/BaseVoiceCommandManager.java [hunk 22]
```diff
@@ -104,172 +109,75 @@ public void dispose() {
...
 
-    private void sendDeleteCurrentVoiceCommands(final CompletionListener listener) {
+    public void setVoiceCommands(List<VoiceCommand> voiceCommands) {
 
-        if (oldVoiceCommands == null || oldVoiceCommands.size() == 0) {
-            if (listener != null) {
-                listener.onComplete(true);
-            }
+        // we actually need voice commands to set.
+        if (voiceCommands == null || voiceCommands.equals(this.voiceCommands)) {
```
[the remark this answers is not in the document]
Comment (RHenigan) [reply base/src/main/java/com/smartdevicelink/managers/screen/menu/BaseVoiceCommandManager.java:134]: Implemented Methods in `VoiceCommand`

File: base/src/main/java/com/smartdevicelink/managers/screen/menu/BaseVoiceCommandManager.java [hunk 23]
```diff
@@ -104,172 +109,75 @@ public void dispose() {
...
+            DebugTool.logInfo(TAG, "Voice commands list was null, returning");
```
[the remark this answers is not in the document]
Comment (RHenigan) [reply base/src/main/java/com/smartdevicelink/managers/screen/menu/BaseVoiceCommandManager.java]: updated log message

File: base/src/main/java/com/smartdevicelink/managers/screen/menu/VoiceCommandUpdateOperation.java [hunk 24]
```diff
@@ -0,0 +1,231 @@
...
+                sendCurrentVoiceCommands(new CompletionListener() {
+                    @Override
+                    public void onComplete(boolean success2) {
+                        if (!success2) {
+                            DebugTool.logError(TAG, "Error sending voice commands");
+                            onFinished();
+                        } else {
+                            DebugTool.logInfo(TAG, "Successfully send voice commands");
+                            onFinished();
+                        }
```
[the remark this answers is not in the document]
Comment (RHenigan) [reply base/src/main/java/com/smartdevicelink/managers/screen/menu/VoiceCommandUpdateOperation.java]: implemented

File: base/src/main/java/com/smartdevicelink/managers/screen/menu/VoiceCommandUpdateOperation.java [hunk 25]
```diff
@@ -0,0 +1,231 @@
...
+                    }
+                }
+            }
+
+            @Override
+            public void onResponse(int correlationId, RPCResponse response) {
+                DeleteCommand foundDeleteCommand = null;
+                for (DeleteCommand deleteCommand : deleteVoiceCommands) {
+                    if (correlationId == deleteCommand.getCorrelationID()) {
+                        foundDeleteCommand = deleteCommand;
```
[the remark this answers is not in the document]
Comment (RHenigan) [reply base/src/main/java/com/smartdevicelink/managers/screen/menu/VoiceCommandUpdateOperation.java:117]: implemented

File: base/src/main/java/com/smartdevicelink/managers/screen/menu/VoiceCommandUpdateOperation.java [hunk 26]
```diff
@@ -0,0 +1,231 @@
...
+                    }
+                }
+            }
+
+            @Override
+            public void onResponse(int correlationId, RPCResponse response) {
+                AddCommand foundAddCommand = null;
+                for (AddCommand addCommand : addCommandsToSend) {
+                    if (correlationId == addCommand.getCorrelationID()) {
+                        foundAddCommand = addCommand;
```
[the remark this answers is not in the document]
Comment (RHenigan) [reply base/src/main/java/com/smartdevicelink/managers/screen/menu/VoiceCommandUpdateOperation.java:191]: implemented

File: android/sdl_android/src/androidTest/java/com/smartdevicelink/managers/screen/menu/VoiceCommandManagerTests.java [hunk 27]
```diff
@@ -34,6 +34,7 @@
```
Comment (bilal-alsharifi) [inline android/sdl_android/src/androidTest/java/com/smartdevicelink/managers/screen/menu/VoiceCommandManagerTests.java:34]: There is an unused import in this class

Comment (RHenigan) [reply android/sdl_android/src/androidTest/java/com/smartdevicelink/managers/screen/menu/VoiceCommandManagerTests.java:34]: updated

File: base/src/main/java/com/smartdevicelink/managers/screen/menu/VoiceCommand.java [hunk 28]
```diff
@@ -124,4 +126,37 @@ void setCommandId(int commandId) {
...
+
+    /**
+     * Used to compile hashcode for VoiceCommand for use to compare in equals method
+     *
+     * @return Custom hashcode of VoiceCommand variables
+     */
+    @Override
+    public int hashCode() {
+        int result = 1;
+        result += ((getCommandId() == 0) ? 0 : Integer.rotateLeft(getCommandId(), 1));
```
[the remark this answers is not in the document]
Comment (RHenigan) [reply base/src/main/java/com/smartdevicelink/managers/screen/menu/VoiceCommand.java]: updated

File: base/src/main/java/com/smartdevicelink/managers/screen/menu/VoiceCommand.java [hunk 29]
```diff
@@ -124,4 +126,37 @@ void setCommandId(int commandId) {
...
+        result += ((getVoiceCommandSelectionListener() == null) ? 0 : Integer.rotateLeft(getVoiceCommandSelectionListener().hashCode(), 2));
```
[the remark this answers is not in the document]
Comment (RHenigan) [reply base/src/main/java/com/smartdevicelink/managers/screen/menu/VoiceCommand.java]: updated

File: base/src/main/java/com/smartdevicelink/managers/screen/menu/VoiceCommand.java [hunk 30]
```diff
@@ -35,6 +35,8 @@
 import androidx.annotation.NonNull;
 import androidx.annotation.Nullable;
 
+import com.smartdevicelink.managers.screen.SoftButtonObject;
```
[the remark this answers is not in the document]
Comment (RHenigan) [reply base/src/main/java/com/smartdevicelink/managers/screen/menu/VoiceCommand.java]: updated

**Review by bilal-alsharifi — APPROVED** [review]

**Pull request merged**